File: Python/Problem1.py
"""
This code provides a solution to two sum problem:

Given an array of integers nums and an integer target, return indices of the two numbers such that they add up to target.

You may assume that each input would have exactly one solution, and you may not use the same element twice.

You can return the answer in any order.

"""
from collections import defaultdict

def findTwoSum(nums, target):
    seen = defaultdict(int) #Keep track of the difference
    n = len(nums)
    for i in range(n):
        remaining = target - nums[i] #difference of the target and each member of the list
        if nums[i] in seen:
            return [i, seen[nums[i]]]
        seen[remaining] = i

def main():
    nums = [3, 3]
    target = 6
    x,y = findTwoSum(nums, target)
    print([x,y])

if __name__ == "__main__":
    main()

File: Cpp/Problem1.cpp
/*
This code takes a vector of integers, and returns the indices of the two numbers
that add up to a given target
*/

#include <iostream>
#include <vector>
using namespace std;

vector <int> twoSum(vector <int> &nums, int);

int main()
{
    int target {11};
    vector <int> nums = {1,3,7,9,2};
    vector indices = twoSum(nums, target);
    cout << indices[0] << '\t' << indices[1] << endl;
}

/*
The brute force solution:
*/

vector <int> twoSum(vector <int> &nums, int target)
{
    vector <int> result;
    for (size_t i = 0; i < nums.size(); i++)
    {
        for (size_t j = i+1; j < nums.size(); j++)
        {
            if (nums[i] + nums[j] == target)
            {
                result.push_back(i);
                result.push_back(j);
                return result;
            }
        }
    }
    
    result.push_back(-1);
    result.push_back(-1);
    return result;
}